### PR TITLE
[Segment Replication] Update force segment replication round to be synchronous

### PR DESCRIPTION
**Pull request by dreamer-89**
### Description
This change fixes the primary relocation path with segrep. This changes:
- Segment replication to happen synchronously. This has side effects of writes happening on older primary and segment replication to timeout due to parallel writes on older primary and round of segment replication from older primary to target.
- Remove `StepListener` listener post above change. This needed corresponding changes in function definitions & unit test changes. 
- Updates existing integration test to reduce flakyness
- Verified newly added tests are not flaky

### Issues Resolved
https://github.com/opensearch-project/OpenSearch/issues/5848

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dreamer-89** [comment]
Converted to draft as it needs more unit test changes and a stablized integration test which verifies delay operations behavior.

**Comment by dreamer-89** [comment]
Tested the change on a 3 data-node 1 master cluster. Tested below use-cases.
1. Large primary shard relocation. With no activity happening during the relocation, the round of segment replication completes in few ms. [Steps gist link](https://gist.github.com/dreamer-89/eb033ddd218cb4c961aec6332c96a70c).
2. Concurrent ingestion during relocation. Noticed that not all document show up post relocation even though all documents were Ack'ed. [Steps gist link](https://gist.github.com/dreamer-89/7cc83bac1ace8bbadb1277358495ef9b)

**Comment by dreamer-89** [comment]
@ashking94 @Bukhtawar @mch2 : Gentle reminder on review.

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java [hunk 1]
```diff
@@ -58,15 +60,20 @@ private void createIndex() {
     @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/5669")
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java:60]: does this still need to be muted?

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java [hunk 2]
```diff
@@ -98,28 +105,29 @@ public void testPrimaryRelocation() throws Exception {
...
-
-        logger.info("--> state {}", state);
-
         assertEquals(
             state.getRoutingNodes().node(state.nodes().resolveNode(newPrimary).getId()).iterator().next().state(),
             ShardRoutingState.STARTED
         );
 
-        final int finalDocCount = initialDocCount;
-        ingestDocs(finalDocCount);
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java:110]: I believe this was added to SegmentReplicationIT for this test case, if its no longer used lets pls remove it.

File: server/src/main/java/org/opensearch/indices/recovery/RemoteRecoveryTargetHandler.java [hunk 3]
```diff
@@ -192,16 +192,17 @@ public void indexTranslogOperations(
...
         final long requestSeqNo = requestSeqNoGenerator.getAndIncrement();
-        final ForceSyncRequest request = new ForceSyncRequest(requestSeqNo, recoveryId, shardId);
-        final Writeable.Reader<TransportResponse.Empty> reader = in -> TransportResponse.Empty.INSTANCE;
-        final ActionListener<TransportResponse.Empty> responseListener = ActionListener.map(listener, r -> null);
-        retryableTransportClient.executeRetryableAction(action, request, responseListener, reader);
+        transportService.submitRequest(
+            targetNode,
+            SegmentReplicationTargetService.Actions.FORCE_SYNC,
+            new ForceSyncRequest(requestSeqNo, recoveryId, shardId),
+            TransportRequestOptions.builder().withTimeout(recoverySettings.internalActionLongTimeout()).build(),
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/recovery/RemoteRecoveryTargetHandler.java:203]: How would a segrep failure be treated vs a network failure?  Are we able to resume recovery after completing phase 1 or would the shard get wiped?

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java:60]: Yes @mch2 as the test occasionally fails due to doc count mismatch towards the end.

File: server/src/main/java/org/opensearch/indices/recovery/RemoteRecoveryTargetHandler.java (hunk 3 above)
Comment (dreamer-89) [reply server/src/main/java/org/opensearch/indices/recovery/RemoteRecoveryTargetHandler.java:203]: A segment replication failure will result in failure of recovery process with shard. So, in event of any exception other than connect exception (mentioned next), the recovery process will be marked failed. I think it is worthwhile to revisit different segrep n/w timeout settings and ensure they are appropriate; without which there are chances of continuous recovery failure due to segment replication. 

On network failure `ConnectTransportException` target sends [REESTABLISH_RECOVERY](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/recovery/PeerRecoveryTargetService.java#L725) request to source which resumes the recovery process using existing handler. If no such handler is found on source, a `PeerRecoveryNotFound` exception is sent which `retry` the recovery from beginning (creates a new recovery request and copying existing recoveryTarget). `Retry` also happens due DelayRecoveryException (source not owner of shard) or IllegalState of shard (IndexNotFoundException/ShardNotFoundException/IllegalIndexShardStateException).

**Comment by dreamer-89** [comment]
> @ashking94 @Bukhtawar @mch2 : Gentle reminder on review.

@ashking94 @Bukhtawar : Ping for review ^

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java (hunk 1 above)
Comment (mch2) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java:60]: https://github.com/opensearch-project/OpenSearch/pull/6015 should help.

**Review by mch2 — APPROVED** [review]
Looks good. Thanks for fixing this!

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java (hunk 1 above)
Comment (dreamer-89) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java]: Thanks @mch2 for fixing flaky tests. The test is not flaky anymore. Ran 100 times without failure.

**Pull request merged**